### PR TITLE
Llama3.2 vision model support

### PR DESCRIPTION
**Pull request by hnyls2002**
## Motivation

- Support encoder-decoder architecture in SGLang.
- Support llama vision model.
- Support CUDA graph and prefix cache for llama vision model

Note that to support CUDA graph for encoder-decoder architecture like llama vision (mllama), we should make `encoder_lens` the part of the cuda graph, as the `full_text_row_masked_out_mask` is decided by `encoder_lens` to skip the text-only req in a mixed batch. 

However, the current cuda graph backend (flashinfer) seems to have trouble handling the mixed batch. So we for now only accept the pure image decoding batch.

Todo in the following PRs:
- [ ] Split attention backends: `sliding_window`, `single_attention`, `cross_attention`
- [ ] Optimize encoder cache locations indexing, and reduce memory usage.

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/pyproject.toml [hunk 1]
```diff
@@ -8,26 +8,46 @@ version = "0.3.2"
...
-    "numpy",
-]
+dependencies = ["requests", "tqdm", "numpy"]
 
 [project.optional-dependencies]
-srt = ["aiohttp", "decord", "fastapi", "hf_transfer", "huggingface_hub", "interegular",
-       "packaging", "pillow", "psutil", "pydantic", "python-multipart",
-       "torch", "torchao", "uvicorn", "uvloop", "zmq",
-       "vllm==0.5.5", "outlines>=0.0.44", "modelscope"]
+srt = [
```
Comment (antinucleon) [inline python/pyproject.toml]: It is a good practice to mark every dependency with explicit version to prevent breaking.

Comment (hnyls2002) [reply python/pyproject.toml]: Sure, but actually in this PR, what I did was just format the dependencies into lines.

File: python/sglang/srt/layers/attention/triton_backend.py [hunk 2]
```diff
@@ -10,6 +10,7 @@
 from sglang.srt.model_executor.forward_batch_info import ForwardBatch
 
 if TYPE_CHECKING:
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/triton_backend.py:12]: add some assert to say cross attention and sliding window attention are not supported.

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 3]
```diff
@@ -41,12 +43,13 @@ def __init__(self, size: int, max_context_len: int, device: str, use_records: bo
...
 
-        if use_records:
-            # records all write operations
-            self.write = self.write_with_records
+    def write(self, indices, values):
+        if self.use_records:
+            return self.write_with_records(indices, values)
         else:
-            self.write = self.write_without_records
+            return self.write_without_records(indices, values)
```
Comment (merrymercy) [inline python/sglang/srt/mem_cache/memory_pool.py]: revert the change? The old code eliminates a `if` during the runtime

File: python/sglang/srt/models/qwen2_vl.py [hunk 4]
```diff
@@ -605,7 +605,11 @@ def forward(
             ]
 
         positions = forward_batch.mrope_positions
-        if image_inputs is None or len(image_inputs) == 0:
+        if (
+            forward_batch.forward_mode.is_decode()
+            or image_inputs is None
```
Comment (merrymercy) [inline python/sglang/srt/models/qwen2_vl.py:610]: why this change?

File: python/sglang/srt/layers/attention/triton_backend.py (hunk 2 above)
Comment (hnyls2002) [reply python/sglang/srt/layers/attention/triton_backend.py:12]: https://github.com/sgl-project/sglang/blob/a1f18757eb796aff0bde9ab454e788fc970bbbd7/python/sglang/srt/model_executor/model_runner.py#L507-L519

File: python/sglang/srt/models/qwen2_vl.py (hunk 4 above)
Comment (hnyls2002) [reply python/sglang/srt/models/qwen2_vl.py:610]: Llama vision requires passing image_inputs in both decode and prefill. However, qwen2 vl does not pass image_inputs when decoding, and it uses this as a judge to process the images or not...

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 5]
```diff
@@ -154,7 +160,7 @@ def get_kv_buffer(self, layer_id: int) -> Tuple[torch.Tensor, torch.Tensor]:
 
     def set_kv_buffer(
         self,
-        layer_id: int,
+        layer: RadixAttention,
```
Comment (hcyz33) [inline python/sglang/srt/mem_cache/memory_pool.py:163]: Why is it necessary to change the data type from int to RadixAttention here?
